common/scala/src/main/scala/org/apache/openwhisk/core/entity/WhiskEntity.scala

```diff
@@ -124,6 +125,16 @@ object WhiskDocumentReader extends DocumentReader {
       case x if x == classOf[WhiskRule]           => WhiskRule.serdes.read(value)
       case _                                      => throw DocumentUnreadable(Messages.corruptedEntity)
     }
+    value.asJsObject.fields.get("entityType") match {
+      case Some(JsString(entityType)) => {
+        if (doc.entityType != entityType) {
+          throw DocumentTypeMismatchException(
+            s"document type ${doc.entityType} did not match expected type ${entityType}.")
+        }
+      }
+      case _ =>
+    }
+    doc
```

Comment:
Warning Scala Nit ahead. You can simplify this further:

```scala

    value.asJsObject.fields.get("entityType").foreach {
      case JsString(entityType) if (doc.entityType != entityType) =>
        throw DocumentTypeMismatchException(s"document type ${doc.entityType} did not match expected type $entityType.")

      case _ =>
    }

```

The first case match will only apply if the value of `entityType` is a string and it matches the value in the `doc` type. Otherwise, `_` matches all others.